### PR TITLE
emacs: handle some arguments correctly for genericBuild and melpaBuild

### PR DESCRIPTION
**Pull request by jian-lin**
## Description of changes

See commit messages for more info.

Closes https://github.com/NixOS/nixpkgs/pull/328400 CC @JaviMerino

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by AndersonTorres — APPROVED** [review]
LGTM

File: pkgs/applications/editors/emacs/build-support/generic.nix [hunk 1]
```diff
@@ -4,7 +4,8 @@
 
 let
   inherit (lib) optionalAttrs getLib;
-  handledArgs = [ "buildInputs" "packageRequires" "meta" ];
+  handledArgs = [ "buildInputs" "packageRequires" "propagatedUserEnvPkgs" "meta" ]
+    ++ lib.optionals (emacs.withNativeCompilation or false) [ "nativeBuildInputs" "postInstall" ];
```
Comment (jian-lin) [inline pkgs/applications/editors/emacs/build-support/generic.nix:8]: a bit ugle

**Comment by JaviMerino** [comment]
Tested that it fixes [the issue with emacsPackages.copilot](https://github.com/NixOS/nixpkgs/pull/327952#issuecomment-2238916949). ✅

File: pkgs/applications/editors/emacs/build-support/generic.nix [hunk 2]
```diff
@@ -51,7 +55,7 @@ stdenv.mkDerivation (finalAttrs: ({
 
   buildInputs = [emacs texinfo] ++ packageRequires ++ buildInputs;
   propagatedBuildInputs = packageRequires;
```
Comment (jian-lin) [inline pkgs/applications/editors/emacs/build-support/generic.nix:57]: For elisp package builders (genericBuild, trivialBuild, elpaBuild, melpaBuild), all attributes can only be in one of two states:
- they are overridden by user-specified ones
- or they are merged with user-specified ones.

I think presumably some attributes should be classified into the second class.  So I change those attributes in this PR and leave others untouched.

I do not think there is a principled way to classify them though.  Any thoughts from people who are familiar with builders in other ecosystems?

**Pull request merged**